// valtio/umd/macro/vite.development.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@babel/helper-module-imports'), require('@babel/types'), require('aslemammad-vite-plugin-macro'), require('babel-plugin-macros')) :
  typeof define === 'function' && define.amd ? define(['exports', '@babel/helper-module-imports', '@babel/types', 'aslemammad-vite-plugin-macro', 'babel-plugin-macros'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.valtioVite = {}, global.helperModuleImports, global.t, global.aslemammadVitePluginMacro, global.babelPluginMacros));
})(this, (function (exports, helperModuleImports, t, aslemammadVitePluginMacro, babelPluginMacros) { 'use strict';

  function _interopNamespace(e) {
    if (e && e.__esModule) return e;
    var n = Object.create(null);
    if (e) {
      Object.keys(e).forEach(function (k) {
        if (k !== 'default') {
          var d = Object.getOwnPropertyDescriptor(e, k);
          Object.defineProperty(n, k, d.get ? d : {
            enumerable: true,
            get: function () { return e[k]; }
          });
        }
      });
    }
    n["default"] = e;
    return Object.freeze(n);
  }

  var t__namespace = /*#__PURE__*/_interopNamespace(t);

  var valtioMacro = aslemammadVitePluginMacro.defineMacro("useProxy").withSignature("<T extends object>(proxyObject: T): void").withHandler(function (ctx) {
    var _args$, _path$parentPath, _path$parentPath2, _path$parentPath2$get;

    var path = ctx.path,
        args = ctx.args;
    var hook = helperModuleImports.addNamed(path, 'useSnapshot', 'valtio');
    var proxy = (_args$ = args[0]) == null ? void 0 : _args$.node;
    if (!t__namespace.isIdentifier(proxy)) throw new babelPluginMacros.MacroError('no proxy object');
    var snap = t__namespace.identifier("valtio_macro_snap_" + proxy.name);
    (_path$parentPath = path.parentPath) == null ? void 0 : _path$parentPath.replaceWith(t__namespace.variableDeclaration('const', [t__namespace.variableDeclarator(snap, t__namespace.callExpression(hook, [proxy]))]));
    var inFunction = 0;
    (_path$parentPath2 = path.parentPath) == null ? void 0 : (_path$parentPath2$get = _path$parentPath2.getFunctionParent()) == null ? void 0 : _path$parentPath2$get.traverse({
      Identifier: function Identifier(p) {
        if (inFunction === 0 && p.node !== proxy && p.node.name === proxy.name) {
          p.node.name = snap.name;
        }
      },
      Function: {
        enter: function enter() {
          ++inFunction;
        },
        exit: function exit() {
          --inFunction;
        }
      }
    });
  });
  function provideValtioMacro() {
    return aslemammadVitePluginMacro.defineMacroProvider({
      id: 'valtio/macro',
      exports: {
        'valtio/macro': {
          macros: [valtioMacro]
        }
      }
    });
  }
  var macroPlugin = aslemammadVitePluginMacro.createMacroPlugin({}).use(provideValtioMacro());

  exports["default"] = macroPlugin;
  exports.provideValtioMacro = provideValtioMacro;
  exports.valtioMacro = valtioMacro;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
